core/src/main/scala/flatgraph/Graph.scala

```diff
@@ -52,7 +52,7 @@ class Graph(val schema: Schema, val storagePathMaybe: Option[Path] = None) exten
 
   /** Note: this included `deleted` nodes! You might want to use `livingNodeCountByKind` instead. */
   private[flatgraph] def nodeCountByKind(kind: Int): Int =
-    if (nodesArray.length <= kind) 0
+    if (nodesArray.length < kind || kind < 0) 0
```

Comment:
The `kind == nodesArray.length` case is off-by-one and should be included in this check -- this ain't julia ;)

I would remove this check entirely: it is clearly an error and an immediate array-oob exception is best for debugging. But if we have the check then we should do it right.

Reply:
whoopsie, thanks for catching
the reason for introducing this check is to provide a more specific error than an array-oob exception which is quite generic

Reply:
Updated - I also change the semantics: we now throw an exception if someone asks for an invalid kind, previously that would just return 0. I'm still undecided which way to go tbh... do you have a preference?